doc/design/block_execution.md

```diff
@@ -0,0 +1,147 @@
+
+However, users' configuration of hardware is simple, maybe only some GPU card ids will be given. In placement pass of compile-time, only source device id and destination device id will bet set.
+
+So, `Executor` need to have a `DeviceContextManager` to initialize some DeviceContext at the very beginning. `DeviceContextManager` maneges all DeviceContexts in all devices.
+
+The `Executor` is defined as follows:
+
+```
+class Executor {
+public:
+  Executor(ProgramDesc*);
+  void Run();
+  
+private:
+  DeviceContextManager_;
+};
+```
+
+
+## Solution
+
+### Compile-time solution
+  
+#### Remove NetOp
+
+At current code base, we have a NetOp defined as follows:
+
+```
+class NetOp : OperatorBase {
+public:
+  void Run(const Scope&, const platform::DeviceContext&);
+  std::vector<std::unique_ptr<OperatorBase>> ops_;
+};
+```
+
+We can compose some basic operators into a big operator, such [FCOp](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/fc_op.cc). And the backward operators generated by [Backward](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/backward.h) is often a NetOp.
+
+There are mainly two breaking point of NetOp unpon the overall design:
+
+- Backward pass should be do at compile-time, not run-time.
+- A NetOp will corresponds only one OpDesc in BlockDesc. We will have a big backward NetOp and run sequentially. It's hard to do potential memory-resuse/kernel-fusion optimization.
+
+We should generate a group of OpDesc at compile-time instead of a NetOp at run-time.
+
+#### Symbolic API
+
+Please refer to the survey [doc](https://github.com/QiJune/Paddle/blob/e90ec7783a1abe7f7627f97559cc46488e41cc7e/doc/design/graph_survey.md) on Computation Graph. Users will write a neural network topology with Symbolic API. And the composition of operators should be done at compile-time in this level too.
+
+#### Unified Pass Interface
+
+An abstract class `Converter` is defined to provide a Unified Pass Interface.
+
+```
+class Converter {
+public:
+  virtual void ApplyPass(ProgramDesc*) = 0;
+};
+```
+
+#### Placement Policy
+
+Placement Policy is designed to set device for every operator. Currently, we only need a simple priority rule to implement the simplest version.
+
+
+
+### Run-time solution
+We will have several class derived from `Executor` to provide different execution strategy. And ProgramDesc will be transformed accordingly.
+
+There are mainly two kinds of `Executor`:
+
+- SimpleExecutor
+
+SimpleExecutor will do little things to ProgramDesc, and just construct operators and variables in order. The operators will be executed sequentially.
+
+
+- DAGExecutor
+
+```
+class DAGExecutor : public Executor {
+public:
+  void Transform();
```

Comment:
I think it's `Converter`'s job to optimize the ProgramDesc, executor's job is to run the optimized ProgramDesc efficiently.

Reply:
Because ProgramDesc uses linear list to represent a topology, it's hard to do further optimization upon ProgramDesc. 
SimpleExecutor will parse ProgramDesc sequentially. DAGExecutor will parse ProgramDesc into Graph, and more optimization will be done upon Graph structure.

Reply:
> DAGExecutor will parse ProgramDesc into Graph, and more **optimization** will be done upon Graph structure.

As I said, I think `Executor` should not do optimization, it's the `Converter`'s job to optimize the ProgramDesc.